include/seqan3/search/kmer_index/shape.hpp

```diff
@@ -0,0 +1,141 @@
+* account when computing the hash value. A `1` therefore translates to a position that is used to compute the hash
+* value.
+*
+* ### Example
+*
+* \include test/snippet/search/kmer_index/shape.cpp
+*
+* \attention 0 < size <= 30
+*
+* \if DEV
+* ### Implementation details
+*
+* The last position indicates whether the shape is ungapped, i.e. if `back()` returns `true` the shape is ungapped,
+* otherwise not.
+* \endif
+*/
+
+class shape : public small_vector<bool, 31>
+{
+public:
+    /*!\name Constructors, destructor and assignment
+     * \{
+     */
+    constexpr shape()                          noexcept = default; //!< Defaulted.
+    constexpr shape(shape const &)             noexcept = default; //!< Defaulted.
+    constexpr shape(shape &&)                  noexcept = default; //!< Defaulted.
+    constexpr shape & operator=(shape const &) noexcept = default; //!< Defaulted.
+    constexpr shape & operator=(shape &&)      noexcept = default; //!< Defaulted.
+    ~shape()                                   noexcept = default; //!< Defaulted.
+
+    /*!\brief Construct an ungapped shape from a given size.
+     *
+     * ### Complexity
+     *
+     * Linear in k.
+     *
+     * ### Exceptions
+     *
+     * No-throw guarantee.
+     *
+     * \attention The size must be in the interval [1, 30]. In Debug mode, an assertion will check this constraint.
+     */
+    constexpr shape(ungapped k) noexcept : small_vector<bool, 31>(k.get(), true)
+    {
+        assert(k.get() <= 30);
+        push_back(true);
+    }
+
+    /*!\brief Construct from a parameter pack.
+     *
+     * ### Complexity
+     *
+     * Linear in the number of arguments.
+     *
+     * ### Exceptions
+     *
+     * No-throw guarantee.
+     *
+     * \attention The size of the pack must be in the interval [1, 30].
+     */
+    template <typename... t>
+    //!\cond
+        requires sizeof...(t) >= 2 && sizeof...(t) <= 30 && (std::Integral<t> && ...)
+    //!\endcond
+    constexpr shape(t... args) noexcept
+    {
+        assert([] (auto... a) constexpr { return ((a == 0 || a == 1) && ...); }(args...)); // Only 0 and 1 allowed
+        bool ungapped{true};
+        for (bool a : { args... })
+        {
+            push_back(a);
+            ungapped &= a;
+        }
+        assert(front() == 1); // First position must be 1, e.g. no 0111 shape
+        assert(back() == 1); // Last position must be 1, e.g. no 1110 shape
+        push_back(ungapped); //!\if DEV \todo replace by std::ranges::all_of once it is constexpr \endif
+    }
+    //!\}
+    
+    /*!\brief Returns the size of the shape, i.e. the k-mer length.
+     * \returns the k-mer length.
+     *
+     * ### Complexity
+     *
+     * Constant.
+     *
+     * ### Exceptions
+     *
+     * No-throw guarantee.
+     */
+    constexpr size_type size() const noexcept
+    {
+        return sz - 1;
+    }
+};
```

Comment:
Hm, maybe I was to eager with integrating the bool here. Because now one can flip a bit via [] and the ungapped bit will be wrong :(

Reply:
You would need to overload the operator[], as well to prevent that.

Reply:
I think I would have to overload quite a bit if I want to do it fail safe... `back()`, `insert()`, `operator[]`, if it's default constructed the last position is not yet the ungapped indicator....
So maybe just keep a bool in the iterator or view for now?

Reply:
Well, it all affects the interface of shape which part of our API, but I guess there is nothing to be done about it now. I would move it to the view.